### ISSUE TITLE
Unexpected behaviour when calling "Map#addLayer" with invalid "before" argument

### ISSUE THREAD

**Issue by iskandari** [mapbox/mapbox-gl-js#3868]
I am using the Mapbox GL [example](https://www.mapbox.com/mapbox-gl-js/example/setstyle/) change a map's style.

When the style is switched to `dark-v9`, `basic-v9`, or `streets-v9`, existing layers reload with the `map.on('style.load', function () {}`  This is expected. 

However, this is not the case with `satellite-v9` - when the satellite style is toggled, the existing layers load behind the style.  Please see my example [code](http://iskandarblue.github.io/mapbox/code/test_parcels.html)

This is unexpected behavior.  Why does this happen and Is there a workaround?

**Comment by averas** [comment]
The problem is that you are providing a <code>before</code> parameter to addLayer referencing a layer (road-primary) that does exist in some of the styles but _not_ in satellite.

One might assume that this would cause an exception, or at the very least fall back to the default behaviour of adding the layer at the end of the layer stack, however, the current implementation:

```javascript
   const index = before ? this._order.indexOf(before) : this._order.length;
   this._order.splice(index, 0, id);
```

.. will actually place the the layer second to last when there is no hit on the before parameter. In your case that means that your layers are added before the last layer in the satellite style, which is the raster layer. Simply removing the "road-primary" parameter will make your code work: http://codepen.io/anon/pen/mONPJm

Although you did not use the API as documented I would argue that the current behaviour is not logical and can be improved (thrown exception or fallback to last as mentioned above).

**Comment by stevage** [comment]
FWIW (and I realise this may not be helpful feedback), I'm not enjoying this change. I'm often running into situations where an `addLayer` is failing because some other layer isn't present. This happens:

- when I'm temporarily switching my base style out for something that loads faster (now my overlay layers don't get added at all)
- when I'm trying out different base styles
- when for some reason my layers didn't load in the order I anticipated
- when my layers, their names, and ordering are all in a bit of flux

Basically, this change is slowing me down and feels pedantic. I would prefer the "before" param to be taken as a hint ("stick the new layer before layer X if you can, otherwise stick it at the end"). That is, the "...or..." option suggested by the OP here:

>Either a hard fail, or add the layer to the end of the stack with a warning that the specified before doesn't exist.
